### PR TITLE
add monikerRangeParser

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

File: src/docfx/moniker/MonikerRangeParser.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        public (List<Error>, IEnumerable<string>) Parse(string rangeString)
+        {
+            var errors = new List<Error>();
+            IEnumerable<string> monikerNames = null;
+            try
+            {
+                var expression = ExpressionCreator.Create(rangeString);
+                monikerNames = expression.Accept(_monikersEvaluator);
+            }
+            catch (MonikerRangeException ex)
```
Comment (yufeih) [inline src/docfx/moniker/MonikerRangeParser.cs]: replace `MonikerRangeException` with `Errors.InvalidMonikerRange().Throw()` ?

File: src/docfx/moniker/MonikerRangeParser.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Docs.Build
+{
+    internal class MonikerRangeParser
```
Comment (yufeih) [inline src/docfx/moniker/MonikerRangeParser.cs]: moving these to `/docfx/lib/moniker` folder?

File: src/docfx/moniker/MonikersEvaluatorVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class EvaluatorWithMonikersVisitor
```
Comment (superyyrrzz) [inline src/docfx/moniker/MonikersEvaluatorVisitor.cs]: >EvaluatorWithMonikersVisitor [](start = 19, length = 28)

align file name with class name

File: src/docfx/moniker/MonikersEvaluatorVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                case ComparatorOperatorType.GreaterThan:
+                    return monikerProductInfo.OrderedProductMonikerNames.Skip(monikerProductInfo.Index + 1);
+                case ComparatorOperatorType.GreaterThanOrEqualTo:
+                    return monikerProductInfo.OrderedProductMonikerNames.Skip(monikerProductInfo.Index);
+                case ComparatorOperatorType.LessThan:
+                    return monikerProductInfo.OrderedProductMonikerNames.Take(monikerProductInfo.Index);
+                case ComparatorOperatorType.LessThanOrEqualTo:
+                    return monikerProductInfo.OrderedProductMonikerNames.Take(monikerProductInfo.Index + 1);
+                default:
+                    return null;
```
Comment (yufeih) [inline src/docfx/moniker/MonikersEvaluatorVisitor.cs]: `return Array.Empty<>()` over `null`, also applies to below

File: src/docfx/moniker/MonikersEvaluatorVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            var errors = new List<Error>();
+            var monikerName = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            var productNameDictionary = new Dictionary<string, List<Moniker>>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var moniker in monikers)
+            {
+                if (monikerName.Contains(moniker.MonikerName))
+                {
+                    errors.Add(Errors.MonikerNameConflict(moniker.MonikerName));
```
Comment (yufeih) [inline src/docfx/moniker/MonikersEvaluatorVisitor.cs]: just throw here

File: src/docfx/moniker/MonikerRangeParser.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/moniker/MonikerRangeParser.cs]: `throw Errors.InvalidMonikerRange` to break execution, we should not proceed with a bad moniker range.

File: src/docfx/moniker/ExpressionCreator.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Docs.Build
+{
+    internal class ExpressionCreator
+    {
+        private static readonly Regex OrSymbolRegex = new Regex(@"^\s*(?<or>\|\|)", RegexOptions.Compiled);
```
Comment (yufeih) [inline src/docfx/moniker/ExpressionCreator.cs]: can we parse these expressions without regex?

Comment (928PJY) [reply src/docfx/moniker/ExpressionCreator.cs]: using string spplit with `||` or comparators?

---
In reply to: [228094545](https://github.com/dotnet/docfx/pull/3529#discussion_r228094545) [](ancestors = 228094545)

**Pull request merged**